### PR TITLE
[NEUT-1547] - Bkp Section Header

### PR DESCRIPTION
**Pull request by gabriellferreira**
https://skyscanner.atlassian.net/browse/NEUT-1547

## Context / Background 
For use within Destination Explorer, and following[ Backpack First](https://skyscanner.atlassian.net/wiki/download/attachments/150700037/Backpack%20First%20-%20SLT%20-%202022-09.pptx?version=1&modificationDate=1664961404873&cacheVersion=1&api=v2), we aim to develop generic versions of the UI components we will require. This ticket covers the [Section Header](https://www.figma.com/file/yN0hFyZlKL0Jwbpi0rEKYT/Backpack-Beta?type=design&node-id=9950-46078&mode=dev) component (which is part of the [Card List](https://www.figma.com/file/yN0hFyZlKL0Jwbpi0rEKYT/Backpack-Beta?type=design&node-id=10448-49801&mode=dev) component)

Screenshot 2023-07-26 at 17 41 40

+ [X] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [X] Component `README.md`
+ [x] Tests

If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)

### CODE REVIEW

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+package net.skyscanner.backpack.compose.utils
+
+import android.content.res.Configuration
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.platform.LocalConfiguration
+
+@Composable
+fun isTablet(): Boolean {
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt:26]: based on figma the tablet mode starts after 513dp

File: docs/compose/SectionHeader/README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+| --- | --- |
+| <img src="https://raw.githubusercontent.com/Skyscanner/backpack-android/main/docs/compose/SectionHeader/screenshots/default.png" alt="SectionHeader component" width="375" /> | <img src="https://raw.githubusercontent.com/Skyscanner/backpack-android/main/docs/compose/SectionHeader/screenshots/default_dm.png" alt="SectionHeader component - dark mode" width="375" /> |
+
+## Installation
+
+Backpack Compose is available through [Maven Central](https://search.maven.org/artifact/net.skyscanner.backpack/backpack-compose). Check the main [Readme](https://github.com/skyscanner/backpack-android#installation) for a complete installation guide.
+
+## Usage
+
+Example of a SectionHeader:
```
Comment (marianeum) [inline docs/compose/SectionHeader/README.md:19]: also here

**Review by marianeum — COMMENTED** [review]
nice one - a few comments

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/BpkSectionHeader.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+fun BpkSectionHeader(
+    title: String,
+    modifier: Modifier = Modifier,
+    description: String? = null,
+    button: BpkSectionHeaderButton? = null,
+) {
+    BpkSectionHeaderImpl(modifier = modifier, title = title, description = description, button = button)
+}
+
+class BpkSectionHeaderButton(
```
[the remark this answers is not in the document]
Comment (marianeum) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/BpkSectionHeader.kt:36]: it's a string in the api table, though and if it'd be a node we couldn't do the tablet/phone differentiation within the component

[the remark this answers is not in the document]
Comment (anambl) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/BpkSectionHeader.kt:36]: what is the tablet/phone differentiation?
I'm not sure why it's a string in the api table (and forgot to update it to node when we got the web contribution). perhaps string makes more sense on apps?

[the remark this answers is not in the document]
Comment (marianeum) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/BpkSectionHeader.kt:36]: We show the button as icon only on phone and text only on tablet/desktop - that should also apply on web

[the remark this answers is not in the document]
Comment (anambl) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/BpkSectionHeader.kt:36]: ah.. right. We do have that illustrated in the examples i.e. mobile is icon only, but not in code. Will chat with Adam to see whether we need to enforce that and also change the web implementation

File: app/src/main/java/net/skyscanner/backpack/demo/compose/SectionHeaderStory.kt [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+import net.skyscanner.backpack.compose.sectionheader.BpkSectionHeaderIcon
+import net.skyscanner.backpack.compose.tokens.BpkSpacing
+import net.skyscanner.backpack.compose.tokens.Heart
+import net.skyscanner.backpack.demo.components.SectionHeaderComponent
+import net.skyscanner.backpack.demo.meta.ComposeStory
+
+@Composable
+@SectionHeaderComponent
+@ComposeStory("Default")
+fun PreViewDefault(modifier: Modifier = Modifier) {
```
[the remark this answers is not in the document]
Comment (gabriellferreira) [reply app/src/main/java/net/skyscanner/backpack/demo/compose/SectionHeaderStory.kt:44]: Adding the scenario without description too.

File: docs/compose/SectionHeader/README.md (hunk 2 above)
Comment (marianeum) [reply docs/compose/SectionHeader/README.md:19]: both should still be `Section Header` (with a space)

File: docs/compose/SectionHeader/README.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
+# SectionHeader
+
+[![Maven Central](https://img.shields.io/maven-central/v/net.skyscanner.backpack/backpack-compose)](https://search.maven.org/artifact/net.skyscanner.backpack/backpack-compose)
+[![Class reference](https://img.shields.io/badge/Class%20reference-Android-blue)](https://backpack.github.io/android/backpack-compose/net.skyscanner.backpack.compose.sectionheader)
+[![Source code](https://img.shields.io/badge/Source%20code-GitHub-lightgrey)](https://github.com/Skyscanner/backpack-android/tree/main/backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader)
+
+## Default
```
Comment (marianeum) [inline docs/compose/SectionHeader/README.md:7]: If you keep having multiple stories please include all screenshots here (and generate the screenshots please)

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    val configuration = LocalConfiguration.current
+    return if (configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
```
[the remark this answers is not in the document]
Comment (gabriellferreira) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt:28]: 🤔🤔🤔🤔
does this represent portrait/landscape? https://www.figma.com/file/yN0hFyZlKL0Jwbpi0rEKYT/Backpack-Beta?type=design&node-id=17606-55889&mode=dev

Under the small tablet section, we have a vertical example and a horizontal one. 
Screenshot 2023-07-26 at 17 29 34

This is a honest question, my first time working with bpk figma file 😄

File: app/src/main/java/net/skyscanner/backpack/demo/compose/SectionHeaderStory.kt [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+import net.skyscanner.backpack.compose.tokens.Heart
+import net.skyscanner.backpack.demo.R
+import net.skyscanner.backpack.demo.components.SectionHeaderComponent
+import net.skyscanner.backpack.demo.meta.ComposeStory
+import net.skyscanner.backpack.toast.BpkToast
+
+@Composable
+@SectionHeaderComponent
+@ComposeStory("Default")
+fun SectionHeaderStory(modifier: Modifier = Modifier) {
```
[the remark this answers is not in the document]
Comment (gabriellferreira) [reply app/src/main/java/net/skyscanner/backpack/demo/compose/SectionHeaderStory.kt:41]: I decided to go with 2 dif stories, one for each type 😄

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (gabriellferreira) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt]: thanks 🚀

File: docs/compose/SectionHeader/README.md (hunk 5 above)
Comment (gabriellferreira) [reply docs/compose/SectionHeader/README.md:7]: done ✅

**Comment by marianeum** [comment]
Good question - generally we don't, but in this case it's worth double checking with design/others in the channel. I _think_ looking at the web one it is included

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                        type = getButtonType(type),
+                    )
+                }
+            }
+        }
+    }
+}
+
+fun getHorizontalArrangement(tablet: Boolean): Arrangement.HorizontalOrVertical {
+    val size = if (tablet) {
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt:100]: Can we use `BpkSpacing.Lg` here (and times 2 for the tablet version)

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+                    color = getTextColor(type),
+                )
+            }
+        }
+        button?.let {
+            Row {
+                if (isTablet) {
+                    BpkButton(
+                        text = it.text,
+                        onClick = { it.onClick.invoke() },
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt:83]: tiny nit: you can just use `it.onClick()`

File: app/src/main/java/net/skyscanner/backpack/demo/compose/SectionHeaderStory.kt [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+                onClick = {},
+            ),
+        )
+    }
+}
+
+@Composable
+@SectionHeaderComponent
+@ComposeStory("OnDark")
+fun SectionHeaderOnDarkStory(modifier: Modifier = Modifier) {
```
Comment (marianeum) [inline app/src/main/java/net/skyscanner/backpack/demo/compose/SectionHeaderStory.kt:67]: can you make the page background dark as well here please?

**Review by marianeum — COMMENTED** [review]
looks good! mainly some tiny comments and let's clarify the margin question otherwise all good

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt [hunk 11]
```diff
@@ -80,14 +81,14 @@ fun BpkSectionHeaderImpl(
                 if (isTablet) {
                     BpkButton(
                         text = it.text,
-                        onClick = { it.onClick.invoke() },
+                        onClick = { it.onClick },
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt:84]: i think you still need the brackets (`onClick()`) OR you can just set `onClick = it.onClick`

**Comment by marianeum** [comment]
@gabriellferreira checked with @anambl and the component should not include the margin

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt (hunk 11 above)
Comment (gabriellferreira) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt]: ops, fixed ✅

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+) {
+    val isTablet = isTablet()
+    Row(
+        modifier = modifier
+            .background(getBackgroundColor(type = type)),
+        horizontalArrangement = getHorizontalArrangement(isTablet),
+        verticalAlignment = Alignment.Top,
+    ) {
+        Column(
+            verticalArrangement = Arrangement.spacedBy(4.dp, Alignment.Top),
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt:59]: `BpkSpacing.Sm`

**Review by marianeum — COMMENTED** [review]
can you just re-generate the docs screenshot please to make sure it's up to date? otherwise all good :shipit:

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    val size = if (tablet) {
+        BpkSpacing.Lg.times(2)
+    } else {
+        BpkSpacing.Lg
+    }
+    return Arrangement.spacedBy(size)
+}
+
+@Composable
+fun getTextColor(type: BpkSectionHeaderType): Color =
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt:109]: Can we make these functions private?

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+
+package net.skyscanner.backpack.compose.utils
+
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.platform.LocalConfiguration
+
+@Composable
+fun isTablet(): Boolean {
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt:25]: And this one as well

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    val configuration = LocalConfiguration.current
+    return configuration.screenWidthDp > 513
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt:27]: Why 513? Can we create a private constant for this?

**Review by bvitaliyg — COMMENTED** [review]
Just few comments

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt (hunk 15 above)
Comment (gabriellferreira) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt:27]: Checking again I don't know if this is still the best approach.
I'm considering going with the same approach we have for the main app (Sky_IsTablet, defining attributes under diff res folders), but quite not sure if there are better solutions under composable.

any opinions here @bvitaliyg @marianeum

Comment (marianeum) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt:27]: I think for now I would keep it private - later on we might want to extend this to not just handle the tablet vs phone but also small vs large tablet case that is defined in figma and expose that to have one source of truth

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+import net.skyscanner.backpack.compose.sectionheader.BpkSectionHeaderType.Default
+import net.skyscanner.backpack.compose.sectionheader.BpkSectionHeaderType.OnDark
+import net.skyscanner.backpack.compose.text.BpkText
+import net.skyscanner.backpack.compose.theme.BpkTheme
+import net.skyscanner.backpack.compose.tokens.ArrowRight
+import net.skyscanner.backpack.compose.tokens.BpkSpacing
+import net.skyscanner.backpack.compose.utils.isTablet
+
+@Composable
+fun BpkSectionHeaderImpl(
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/sectionheader/internal/BpkSectionHeaderImpl.kt:43]: just spotted: can you make this internal please?

File: app/src/main/java/net/skyscanner/backpack/demo/compose/SectionHeaderStory.kt [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+    Column(
+        modifier = modifier
+            .background(BpkTheme.colors.surfaceContrast)
```
Comment (marianeum) [inline app/src/main/java/net/skyscanner/backpack/demo/compose/SectionHeaderStory.kt:72]: can you also add `fillMaxSize` here so it fills the whole screen (and the background takes up that)?

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+ * limitations under the License.
+ */
+
+package net.skyscanner.backpack.compose.utils
+
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.platform.LocalConfiguration
+
+@Composable
+fun isTablet(): Boolean {
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/utils/DeviceUtil.kt:25]: this should also be internal

File: docs/compose/SectionHeader/README.md [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+    BpkSectionHeader(
+        title = stringResource(R.string.section_header_title),
+        description = stringResource(R.string.section_header_description),
+    )
+```
+
+### Section header with a title, description and trailing button.
+
+```Kotlin
+    import net.skyscanner.backpack.compose.sectionheader.BpkSectionHeader
```
Comment (marianeum) [inline docs/compose/SectionHeader/README.md:52]: please also add the `BpkSectionHeaderButton` import where it's used

File: docs/compose/SectionHeader/screenshots/ondark.png
Comment (marianeum) [inline docs/compose/SectionHeader/screenshots/ondark.png:1]: please regenerate the screenshot once the background fills the screen

**Comment by bvitaliyg** [comment]
@gabriellferreira you need to pull the PR, make an empty commit and push it back again to have the CI running as expected

**Review by marianeum — APPROVED** [review]

**Pull request merged**